components/raftstore-v2/src/worker/cleanup/compact.rs

```diff
@@ -0,0 +1,350 @@
+// Copyright 2023 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::{
+    error::Error as StdError,
+    fmt::{self, Display, Formatter},
+};
+
+use engine_traits::{KvEngine, RangeStats, TabletRegistry, CF_WRITE};
+use fail::fail_point;
+use keys::{DATA_MAX_KEY, DATA_MIN_KEY};
+use slog::{error, info, warn, Logger};
+use thiserror::Error;
+use tikv_util::{box_try, worker::Runnable};
+
+pub enum Task {
+    CheckAndCompact {
+        // Column families need to compact
+        cf_names: Vec<String>,
+        region_ids: Vec<u64>,
+        // The minimum RocksDB tombstones a range that need compacting has
+        tombstones_num_threshold: u64,
+        tombstones_percent_threshold: u64,
+        redundant_rows_threshold: u64,
+        redundant_rows_percent_threshold: u64,
+    },
+}
+
+impl Display for Task {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match *self {
+            Task::CheckAndCompact {
+                ref cf_names,
+                ref region_ids,
+                tombstones_num_threshold,
+                tombstones_percent_threshold,
+                redundant_rows_threshold,
+                redundant_rows_percent_threshold,
+            } => f
+                .debug_struct("CheckAndCompact")
+                .field("cf_names", cf_names)
+                .field("regions", region_ids)
+                .field("tombstones_num_threshold", &tombstones_num_threshold)
+                .field(
+                    "tombstones_percent_threshold",
+                    &tombstones_percent_threshold,
+                )
+                .field("redundant_rows_threshold", &redundant_rows_threshold)
+                .field(
+                    "redundant_rows_percent_threshold",
+                    &redundant_rows_percent_threshold,
+                )
+                .finish(),
+        }
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum Error {
+    #[error("compact failed {0:?}")]
+    Other(#[from] Box<dyn StdError + Sync + Send>),
+}
+
+pub struct Runner<E> {
+    logger: Logger,
+    tablet_registry: TabletRegistry<E>,
+}
+
+impl<E> Runner<E>
+where
+    E: KvEngine,
+{
+    pub fn new(tablet_registry: TabletRegistry<E>, logger: Logger) -> Runner<E> {
+        Runner {
+            logger,
+            tablet_registry,
+        }
+    }
+}
+
+impl<E> Runnable for Runner<E>
+where
+    E: KvEngine,
+{
+    type Task = Task;
+
+    fn run(&mut self, task: Self::Task) {
+        match task {
+            Task::CheckAndCompact {
+                cf_names,
```

Comment:
how about define new struct to  make these parameter easy.

Reply:
Good suggestion.